src/services/chain/messages/kava/evmutil/erc20Conversion.ts

```diff
@@ -0,0 +1,238 @@
+{
+  name = 'erc20Convert';
+  description =
+    'Converts an ERC20 asset to a cosmos Coin or cosmos coin to an ERC20 asset';
+  chainType = ChainType.COSMOS;
+  operationType = OperationType.TRANSACTION;
+  walletMustMatchChainID = true;
+
+  needsWallet = [WalletTypes.METAMASK];
+
+  /**
+   * Parameter definitions for the message.
+   * Used for validation and OpenAI tool generation.
+   */
+  parameters = [
+    {
+      name: 'chainName',
+      type: 'string',
+      description: `name of the chain the user is interacting with, if not specified by the user assume ${ChainNames.KAVA_COSMOS}`,
+      enum: [...Object.keys(chainRegistry[ChainType.COSMOS])],
+      required: true,
+    },
+    {
+      name: 'amount',
+      type: 'string',
+      description: 'Amount to convert',
+      required: true,
+    },
+    {
+      name: 'denom',
+      type: 'string',
+      description: 'Token denomination',
+      required: true,
+    },
+    {
+      name: 'direction',
+      type: 'string',
+      description:
+        'direction of the conversion either from cosmos Coin to erc20 or erc20 to cosmos Coin',
+      enum: ['coinToERC20', 'ERC20ToCoin'],
+      required: true,
+    },
+  ];
+
+  inProgressComponent() {
+    return InProgressTxDisplay;
+  }
+
+  validate(params: ERC20ConvertParams, walletStore: WalletStore): boolean {
+    if (!walletStore.getSnapshot().isWalletConnected) {
+      throw new Error('please connect to a compatible wallet');
+    }
+
+    if (Array.isArray(this.needsWallet)) {
+      if (!this.needsWallet.includes(walletStore.getSnapshot().walletType)) {
+        throw new Error('please connect to a compatible wallet');
+      }
+    }
+
+    if (!chainRegistry[this.chainType][params.chainName]) {
+      throw new Error(`unknown chain name ${params.chainName}`);
+    }
+
+    const { amount, denom } = params;
+
+    if (Number(amount) <= 0) {
+      throw new Error(`amount must be greater than zero`);
+    }
+
+    const chainInfo = chainRegistry[this.chainType][params.chainName];
+
+    if (chainInfo.chainType !== ChainType.COSMOS) {
+      throw new Error('chain Type must be Cosmos for this operation');
+    }
+
+    if (
+      !chainInfo.evmChainName ||
+      !chainRegistry[ChainType.EVM][chainInfo.evmChainName ?? '']
+    ) {
+      throw new Error(
+        `cosmos chain ${chainInfo.name} must be linked to an EVM chain`,
+      );
+    }
+
+    const { erc20Contracts } = chainRegistry[ChainType.EVM][
+      chainInfo.evmChainName
+    ] as EVMChainConfig;
+
+    if (
+      !getERC20Record(denom, erc20Contracts) ||
+      !getCoinRecord(denom, chainInfo.denoms)
+    ) {
+      throw new Error(
+        `failed to find contract address or coin record for ${denom}`,
+      );
+    }
+
+    if (
+      params.direction !== 'ERC20ToCoin' &&
+      params.direction !== 'coinToERC20'
+    ) {
+      throw new Error(`unknown conversion direction ${params.direction}`);
+    }
+
+    return true;
```

Comment:
Before returning true, we should also validate that the user has enough balances for the tx to succeed.

![Screenshot 2025-01-31 at 12 19 53 PM](https://github.com/user-attachments/assets/3d46ee73-17e4-4760-9f77-193f247945a7)

https://github.com/Kava-Labs/oros/blob/eba1154abb244f89e0d59b3abe84e52f7cfe28f6/src/services/chain/messages/evm/transfer.ts#L150-L152